src/EditorFeatures/CSharp/MisplacedUsings/MisplacedUsingsCodeFixProvider.SourceMap.cs

```diff
@@ -0,0 +1,281 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.CodeAnalysis.CSharp.MisplacedUsings
+{
+    /// <summary>
+    /// Implements a code fix for all misplaced using statements.
+    /// </summary>
+    internal partial class MisplacedUsingsCodeFixProvider
+    {
+        /// <summary>
+        /// Contains a map of the different regions of a source file.
```

Comment:
what is this needed for?

Reply:
The ordering mechanism allows using directives to be grouped into regions, within with sort order is preserved. Ordering within one group is not checked with respect to other groups.

Reply:
i can't make heads or tails of htat :)

What does ordering/grouping/etc. have to do with moving usings from outside a namespace to inside (or vice versa)?

Reply:
Gotcha.  That's currently definitely *not* the Roslyn way.  The Roslyn way is to have a general command called "organize usings" and individual operatoins internally which build that up.  i.e. sorting, grouping, removing unused.

My expectation would be that we would:
1. add the individual functionality to 'move' usings to the right location.
2. update 'organize usings' to respect that.  (note: this wasn't done in this PR, but it really need to be.
3. Have an analyzer that checks if your usings are *not organized* and offers to fully organize them.

'3' would effectively be equivalent to what you were saying about SyleCopAnalyzers.  Except it would be about all the using-organizaiton options, and not just aobut "are the usings in the right container".

@JoeRobich I would basically expect all this given that we already have the other style options, and we already have the "organize" command.  Right now, this PR feels disjointed because thing aren't fitting together in a common way.